src/pgduckdb_metadata_cache.cpp

```diff
@@ -1,45 +1,157 @@
 extern "C" {
 #include "postgres.h"
 
+#include "access/htup_details.h"
+#include "catalog/dependency.h"
+#include "catalog/namespace.h"
+#include "catalog/pg_type.h"
+#include "catalog/pg_proc.h"
 #include "commands/extension.h"
+#include "nodes/bitmapset.h"
+#include "utils/builtins.h"
+#include "utils/catcache.h"
 #include "utils/inval.h"
+#include "utils/memutils.h"
 #include "utils/syscache.h"
 }
 
+#include "pgduckdb/vendor/pg_list.hpp"
+
 namespace pgduckdb {
 struct {
+	/*
+	 * Does the cache contain valid data, i.e. is it initialized? Or is it
+	 * stale and does it need to be refreshed? If this is false none of the
+	 * other fields should be read.
+	 */
 	bool valid;
+	/*
+	 * Is the pg_duckdb extension installed? If this is false all the other
+	 * fields (except valid) should be ignored. It is totally fine to have
+	 * valid=true and installed=false, this happens when the extension is not
+	 * installed and we cached that information.
+	 */
 	bool installed;
+	/* The Postgres OID of the pg_duckdb extension. */
+	Oid extension_oid;
+	/* A list of Postgres OIDs of functions that can only be executed by DuckDB */
+	List *duckdb_only_functions;
 } cache = {};
 
 bool callback_is_configured = false;
+/* The hash value of the "duckdb" schema name */
 uint32 schema_hash_value;
 
+/*
+ * This function is called for every pg_namespace tuple that is invalidated. We
+ * only invalidate our cache if the "duckdb" schema is invalidated, because
+ * that means that the extension was created or dropped (see comment in
+ * IsExtensionRegistered for details).
+ */
 static void
 InvalidateCaches(Datum arg, int cache_id, uint32 hash_value) {
 	if (hash_value != schema_hash_value) {
 		return;
 	}
+	if (!cache.valid) {
+		return;
+	}
 	cache.valid = false;
+	list_free(cache.duckdb_only_functions);
+}
+
+/*
+ * Builds the list of Postgres OIDs of functions that can only be executed by
+ * DuckDB. The resulting list is stored in cache.duckdb_only_functions.
+ */
+static void
+BuildDuckdbOnlyFunctions() {
+	/* This function should only be called during cache initialization */
+	Assert(!cache.valid);
+	Assert(!cache.duckdb_only_functions);
+	Assert(cache.extension_oid != InvalidOid);
+
+	/*
+	 * We search the system cache for functions with these specific names. It's
+	 * possible that other functions with the same also exist, so we check if
+	 * each of the found functions is actually part of our extension before
+	 * caching its OID as a DuckDB-only function.
+	 */
+	const char *function_names[] = {"read_parquet", "read_csv", "iceberg_scan"};
```

Comment:
Is there a way to fetch all functions that are installed via extension - rather than keep list hardcoded here?

Reply:
just quick note, functions are used only in duckdb live in `public` schema while this function can be found in `duckdb`. But i agree that this is something to be reconsider in future if there need for it.